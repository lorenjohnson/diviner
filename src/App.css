.App {
  display: flex;
  justify-content: space-between;
  min-height: calc(100vh - 10em);
  width: calc(100vw - 10em);
  padding: 5em;
  background-color: black;
  color: antiquewhite;
  box-shadow: inset 0px 0px 100px #44a4e2;
  font-family: Georgia, 'Times New Roman', Times, serif;
}

.left-sidebar {
  width: fit-content;
}

.draw-card {
  box-shadow: -10px 10px 2px 0px antiquewhite;
  border: solid 0.7px #d3d3d3bd;
  filter: opacity(59%);
  cursor: pointer;
  transition: filter 1s ease;
}

.draw-card:hover {
  filter: unset;
}

.drawn-cards:hover .card {
  opacity: 0.5;
}

.drawn-cards .card:not(.face-down):hover {
  opacity: 1;
  box-shadow: 3px 3px 0px 0px rgba(245, 243, 243, 0.61);
}

.draw-card .card-front:hover:after {
  content: 'Draw';
  text-shadow: 4px 4px 20px #44a4e2;
  position: absolute;
  margin-top: 1rem;
  left: 3.6em;
  display: flex;
  font-size: 3em;
  color: white;
}

.drawn-cards {
  display: flex;
  margin-right: auto;
  flex-wrap: wrap;
  max-height: fit-content;
}

.right-sidebar {
  min-width: 20rem;
  max-width: 20rem;
}

.diviner-introduction {
  animation: slide-over 1s ease;
}

.card-detail {
  animation: slide-over 1s ease;
}

@keyframes slide-over {
  0% {
      opacity: 0;
      transform: translateX(20px);
  }
  100% {
      opacity: 1;
      transform: translateX(0);
  }
}