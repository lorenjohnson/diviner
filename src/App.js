import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  DECK_PROP_TYPE,
  DEFAULT_DECK,
  getRandomCard,
  getCard
} from 'util/deck'
import Card from 'components/Card'
import CardDetail from 'components/CardDetail'
import './App.css'

export function drawCard (state, props) {
  const { drawnCards } = state
  const { currentDeck, drawCardFunction } = props
  const drawnCard = drawCardFunction(currentDeck.cards, drawnCards)

  if (!drawnCard) return null

  return {
    drawnCards: [...drawnCards, drawnCard]
  }
}

export default class App extends Component {
  static propTypes = {
    currentDeck: DECK_PROP_TYPE,
    drawCardFunction: PropTypes.func.isRequired
  }

  static defaultProps = {
    currentDeck: DEFAULT_DECK,
    drawCardFunction: getRandomCard
  }

  state = {
    drawnCards: [],
    detailCard: undefined
  }

  drawCard = () => this.setState(drawCard)

  showCardDetail = cardId => this.setState(state => ({
    detailCard: getCard(cardId, this.props.currentDeck.cards)
  }))

  hideCardDetail = () => this.setState(state => ({
    detailCard: undefined
  }))

  render () {
    const { currentDeck: { backImageURL } } = this.props
    const { drawnCards, detailCard } = this.state

    return (
      <div className='App'>
        <div className='left-sidebar'>
          <div className='deck'>
            <Card frontImageURL={backImageURL}
              backImageURL={backImageURL}
              className='draw-card'
              onClick={this.drawCard} />
          </div>
        </div>
        <div className='drawn-cards'>
          {drawnCards.map(({ imageURL, cardId }) =>
            <Card cardId={cardId}
              frontImageURL={imageURL}
              backImageURL={backImageURL}
              onMouseOver={this.showCardDetail}
              onMouseOut={this.hideCardDetail}
              key={cardId} />
          )}
        </div>
        <div className='right-sidebar'>
          {detailCard && <CardDetail className='card-detail' card={detailCard} />}
          {!detailCard && <DivinerIntro />}
        </div>
      </div>
    )
  }
}

function DivinerIntro () {
  return <div className='diviner-introduction'>
    <h2>Welcome to Diviner</h2>
    <h3>"For what we don't know..."</h3>
    <p>
      Click on the card to the right to pick a random card from the <a href='https://www.wikiwand.com/en/Rider-Waite_tarot_deck' target='_blank'>Rider-Waite Tarot deck</a>.
    </p>
    <p>
      You can turn over as many cards as you like and if you roll-over any of the turned-over cards you'll see a more in-depth description of the card.
    </p>
  </div>
}
