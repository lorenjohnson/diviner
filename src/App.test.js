import React from 'react'
import ReactTestRenderer from 'react-test-renderer'
import App, { drawCard } from './App'

// TODO:
// test detail card vs description switch
// test showCardDetail?

describe('App', () => {
  it('renders without crashing', () => {
    ReactTestRenderer.create(<App />)
  })

  it('draws and renders drawn cards', () => {
    const testDeck = {
      cards: [
        { imageURL: '/3', cardId: '3' },
        { imageURL: '/4', cardId: '4' }
      ],
      backImageURL: '/backimage.png'
    }
    const props = {
      currentDeck: testDeck,
      drawCardFunction: jest.fn((c, d) => c[d.length])
    }
    const component = ReactTestRenderer.create(<App {...props} />)

    component.getInstance().drawCard({}, props)
    component.getInstance().drawCard({}, props)
    component.getInstance().drawCard({}, props)

    expect(props.drawCardFunction).toHaveBeenCalledTimes(3)
    expect(component.toJSON()).toMatchSnapshot()
  })
})

test('drawCard', () => {
  const state = {
    drawnCards: ['1', '2']
  }
  const props = {
    currentDeck: {
      cards: ['3', '4', '5']
    },
    drawCardFunction: jest.fn((c, d) => c[0])
  }
  const expectedState = {
    drawnCards: ['1', '2', '3']
  }

  expect(drawCard(state, props)).toEqual(expectedState)
})
