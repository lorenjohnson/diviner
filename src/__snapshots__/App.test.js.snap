// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App draws and renders drawn cards 1`] = `
<div
  className="App"
>
  <div
    className="left-sidebar"
  >
    <div
      className="deck"
    >
      <div
        className="draw-card card"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="card-back"
        >
          <img
            alt="/backimage.png"
            src="/backimage.png"
          />
        </div>
        <div
          className="card-front"
        >
          <img
            alt="/backimage.png"
            src="/backimage.png"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="drawn-cards"
  >
    <div
      className="card"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="card-back"
      >
        <img
          alt="/backimage.png"
          src="/backimage.png"
        />
      </div>
      <div
        className="card-front"
      >
        <img
          alt="/3"
          src="/3"
        />
      </div>
    </div>
    <div
      className="card"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="card-back"
      >
        <img
          alt="/backimage.png"
          src="/backimage.png"
        />
      </div>
      <div
        className="card-front"
      >
        <img
          alt="/4"
          src="/4"
        />
      </div>
    </div>
  </div>
  <div
    className="right-sidebar"
  >
    <div
      className="diviner-introduction"
    >
      <h2>
        Welcome to Diviner
      </h2>
      <h3>
        "For what we don't know..."
      </h3>
      <p>
        Click on the card to the right to pick a random card from the 
        <a
          href="https://www.wikiwand.com/en/Rider-Waite_tarot_deck"
          target="_blank"
        >
          Rider-Waite Tarot deck
        </a>
        .
      </p>
      <p>
        You can turn over as many cards as you like and if you roll-over any of the turned-over cards you'll see a more in-depth description of the card.
      </p>
    </div>
  </div>
</div>
`;

exports[`App shows detail card if set 1`] = `
<div
  className="App"
>
  <div
    className="left-sidebar"
  >
    <div
      className="deck"
    >
      <div
        className="draw-card card"
        onClick={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <div
          className="card-back"
        >
          <img
            alt="/backimage.png"
            src="/backimage.png"
          />
        </div>
        <div
          className="card-front"
        >
          <img
            alt="/backimage.png"
            src="/backimage.png"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="drawn-cards"
  />
  <div
    className="right-sidebar"
  >
    <div
      className="card-detail"
    >
      <h1 />
      <p />
      <h2>
        Meaning
      </h2>
      <p />
    </div>
  </div>
</div>
`;
